README

             decompyle -- A Python byte-code decompiler
                          version 2.4, 02 Jan 2006

Introduction
------------

'decompyle' converts Python byte-code back into equivalent Python
source. It accepts byte-code from any Python version starting with 1.5
up to 2.4.

The generated source is very readable: docstrings, lists, tuples and
hashes get pretty-printed.

'decompyle' may also verify the equivalence of the generated source by
by compiling it and comparing both byte-codes.

'decompyle' is based on John Aycock's generic small languages compiler
'spark' (http://www.csr.uvic.ca/~aycock/python/) and his prior work on
'decompyle'.

Additional note (3 July 2004, Ben Burton):

    The original website from which this software was obtained is no longer
    available.  It has now become a commercial decompilation service, with
    no software available for download.

    Any developers seeking to make alterations or enhancements to this code
    should therefore consider these debian packages an appropriate starting
    point.

Features
--------

  * decompyles Python byte-code into equivalent Python source

  * decompyles byte-code from Python versions 1.5 up to 2.4

  * pretty-prints docstrings, hashes, lists and tuples
  
  * supports an option for verifing the generated Python source is
    equivalent to the byte-code. This is done by compiling the
    generated source and comparing both byte-codes.

  * decompyles and successfully verifies 100% of the Python 1.5, 2.0,
    2.1 and 2.2 library

  * decompyles and successfully verifies 100% of the Python 1.5
    library, including lib-stdwin, lib-tk, gnome and gtk.

  * reads directly from .pyc/.pyo files, bulk-decompyle whole
    directories

  * output may be written to file, a directory or to stdout

  * option for including byte-code disassembly into generated source

  * If translation fails, the part causing the problem is output. This
    helps improofing 'decompyle'. If this happens to you, please
    contact the author.

  For a list of changes please refer to the 'CHANGES' file.


Requirements
------------

'decompyle' requires Python 2.2 or later.


Installation
------------

You may either create a RPM and install this, or install directly from
the source distribution.

Creating RPMS:

  python setup.py bdist_rpm

  If you need to force the python interpreter to eg. pyton2:
     python2 setup.py bdist_rpm --python=python2


Installation from the source distribution:

     python setup.py install

   To install to a user's home-dir:
     python setup.py install --home=<dir>

   To install to another prefix (eg. /usr/local)
     python setup.py install --prefix=/usr/local

  If you need to force the python interpreter to eg. pyton2:
     python2 setup.py install

  For more information on 'Installing Python Modules' please refer to
     http://www.python.org/doc/current/inst/inst.html


Usage
-----

decompyle -h		prints short usage
decompyle --help	prints long usage


Known Bugs/Restrictions
-----------------------

* EXTENDED_ARG token is untested (this is a new token for Python 2.0
  which is used only if many items exist within a code object).

* Verifying decompyled source with optizimzed byte code (.pyo) when
  running without optimizations (option '-O' not given) fails in most
  cases. Same is true for vis-a-versa. This is due to the fact that
  Python generated different bytecode depending on option '-O'.


* Python 2.2 generated different byte-code than prior version for the
  same source. This is due the intruduction of iterators. Currently
  'decompyle' fails verifying this source if the byte-code was
  generated by an older version of Python.


CHANGES
release 2.4 (Dan Pascu)
  - Replaced the way code structures are identified by the parser.
    Previously, the scanner introduced some COME_FROM entries in the
    dissasembly output to mark all the destinations of jump instructions.
    Using these COME_FROM labels the parser was then able to identify the
    code structures (if tests, while loops, etc). Up to python-2.3 this was
    possible because the code structures were clearly defined and jump
    targets were always to the same points in a given strcuture making it
    easy to identify the structure. Python 2.3 however introduced optimized
    jumps to increase code performance. In the previous version of decompyle
    (2.3) we used a technique to identify the code structures and then used
    these structures to determine where the jump targets would have been if
    not optimized. Using this information we then added COME_FROM labels at
    the points where they would have been if not optimized, thus emulating
    the way decompyle worked with versions before python 2.3. However with
    the introduction of even more optimizations in python 2.4 this technique
    no longer works. Not only the jump targets are no longer an effective
    mean for the parser to identify the code structures, but also trying to
    emulate the old way things were solved when it clearly no longer works
    is not the right solution. To solve this issue, the code to identify the
    structures that we had developed in version 2.3, was used to add real
    start/end points for strcuture identification, instead of the COME_FROM
    labels. Now these new start/end labels are used by the parser to more
    precisely identify the structures and the COME_FROM labels were removed
    completely. The scanner is responsible to identify these code structures
    and use any knowledge of optimizations that python applies to determine
    the start/end points of any structure and then mark them with certain
    keywords that are understood by the parser.
  - Correctly identify certain `while 1' structures that were not
    recognized in the previous version.
  - Added support for new byte code constructs used by python 2.4

release 2.3.2
  - tidied up copyright and changelog information for releases 2.3 and later

release 2.3.1 (Dan Pascu)
  - implemented a structure detection technique that fixes problems with
    optimised jumps in Python >= 2.3. In the previous release (decompyle 2.3),
    these problems meant that some files were incorrectly decompiled and
    others could not be decompiled at all. With this new structure detection
    technique, thorough testing over the standard python libraries suggests
    that decompyle 2.3.1 can handle everything that decompyle 2.2beta1 could,
    plus new Python 2.3 bytecodes and constructs.

release 2.3 (Dan Pascu)
  - support for Python 2.3 added
  - use the marshal and disassembly code from their respective python
    versions, so that decompyle can manipulate bytecode independently
    of the interpreter that runs decompyle itself (for example it can
    decompile python2.3 bytecode even when running under python2.2)

release 2.2beta1 (hartmut Goebel)
  - support for Python 1.5 up to Python 2.2
  - no longer requires to be run with the Python interpreter version
    which generated the byte-code.
  - requires Python 2.2
  - pretty-prints docstrings, hashes, lists and tuples
  - decompyle is now a script and a package
  - added emacs mode-hint and tab-width for each file output
  - enhanced test suite: more test patterns, .pyc/.pyo included
  - avoids unnecessary 'global' statements
  - still untested: EXTENDED_ARG

  internal changes:
  - major code overhoul: splitted into several modules, clean-ups
  - use a list of valid magics instead of the single one from imp.py
  - uses copies of 'dis.py' for every supported version. This ensures
    correct disassemling of the byte-code.
  - use a single Walker and a single Parser, thus saving time and memory
  - use augmented assign and 'print >>' internally
  - optimized 'Walker.engine', the main part of code generation

release 0.6.0: (hartmut Goebel)
  - extended print (Python 2.0)
  - extended import (Python 2.0) (may not cover all cases)
  - augmented assign (Python 2.0) (may not cover all cases)
  - list comprehensions (Python 2.0)
  - equivalent for 'apply' (Python 1.6)
  - if .. elif .. else are now nested as expected
  - assert test, data
  - unpack list corrected (was the same as unpack tuple)
  - fixed unpack tuple (trailing semicolon was missing)
  - major speed up :-)
  - reduced memory usage (pre-alpha-0.5 has increased it a lot)
  - still missing: EXTENDED_ARG

pre-alpha-0.5: (hartmut Goebel)
  - *args, **kwargs
  - global
  - formal tuple parameters (eg. def a(self, (x,y,z)) )
  - actual lambda parameters (eg.  X(lambda z: z**2) )
  - remove last 'return None' in procedures
  - remove last 'return locals()' in class definitions
  - docstrings

pre-alpha-0.4: (hartmut Goebel)
  - assert
  - try/except/finally
  - parentheses in expressions
  - nested expressions
  - extracted dissassemble() from module dis and
    removed ugly redirect of stdout, thus saved a lot of
    ugly code and a lot of memory

pre-alpha-0.3: (hartmut Goebel)
  - keyword arguments
  - some boolean expressions
  - and/or
  - complex conditions in if/while
  - read byte-code from .pyc without importing
  - access to the body of classes and modules
  - class and function definitions
  - a = b = c = xxx

pre-alpha-0.1 -> pre-alpha-0.2:
  -  SET_LINENO filtered out in lexer now
  -  added support for subscripts (just for Christian Tismer :-)
  -  fixed bug with handling of BUILD_{LIST,TUPLE} & CALL_FUNCTION
  -  dict-building support
  -  comparison support
  -  exec support
  -  del support
  -  pass support
  -  slice support
  -  no more extraneous (albeit legal) commas
  -  finally, it excepts try [sic] but not all 42 variations of it


decompyle/Parser.py
#  Copyright (c) 1999 John Aycock
#  Copyright (c) 2000-2002 by hartmut Goebel <hartmut@goebel.noris.de>
#  Copyright (c) 2005 by Dan Pascu <dan@windowmaker.org>
#
#  See main module for license.
#

__all__ = ['parse', 'AST', 'ParserError', 'Parser']

from spark import GenericASTBuilder, GenericASTMatcher
import string, exceptions, sys
from UserList import UserList

from Scanner import Token

class AST(UserList):
	def __init__(self, type, kids=[]):
		self.type = intern(type)
		UserList.__init__(self, kids)

	def __getslice__(self, low, high):	return self.data[low:high]
	def __eq__(self, o):
		if isinstance(o, AST):
			return self.type == o.type \
			       and UserList.__eq__(self, o)
		else:
			return self.type == o

	def __hash__(self):			return hash(self.type)

	def __repr__(self, indent=''):
		rv = str(self.type)
		for k in self:
			rv = rv + '\n' + string.replace(str(k), '\n', '\n   ')
		return rv


class ParserError(Exception):
    def __init__(self, token, offset):
        self.token = token
        self.offset = offset

    def __str__(self):
        return "Syntax error at or near `%r' token at offset %s" % \
               (self.token, self.offset)
    

class Parser(GenericASTBuilder):
	def __init__(self):
		GenericASTBuilder.__init__(self, AST, 'stmts')
		self.customized = {}

	def cleanup(self):
		"""
		Remove recursive references to allow garbage
		collector to collect this object.
		"""
		for dict in (self.rule2func, self.rules, self.rule2name, self.first):
			for i in dict.keys():
				dict[i] = None
		for i in dir(self):
			setattr(self, i, None)

	def error(self, token):
            raise ParserError(token, token.offset)

	def typestring(self, token):
		return token.type
	
	def p_funcdef(self, args):
		'''
		stmt ::= funcdef
		funcdef ::= mkfunc designator
		load_closure ::= load_closure LOAD_CLOSURE
		load_closure ::= LOAD_CLOSURE
		'''

	def p_list_comprehension(self, args):
		'''
		expr ::= list_compr
		list_compr ::= BUILD_LIST_0 DUP_TOP _load_attr
				designator list_iter del_stmt

		list_iter ::= list_for
		list_iter ::= list_if
		list_iter ::= lc_body

        _load_attr ::= LOAD_ATTR
        _load_attr ::=

		_lcfor ::= GET_ITER LIST_COMPREHENSION_START FOR_ITER
		_lcfor ::= LOAD_CONST FOR_LOOP
		_lcfor2 ::= GET_ITER FOR_ITER
		_lcfor2 ::= LOAD_CONST FOR_LOOP

		list_for ::= expr _lcfor designator list_iter
				LIST_COMPREHENSION_END JUMP_ABSOLUTE

		list_for ::= expr _lcfor2 designator list_iter
				JUMP_ABSOLUTE

		list_if ::= expr condjmp IF_THEN_START list_iter
				IF_THEN_END _jump POP_TOP IF_ELSE_START IF_ELSE_END

		lc_body ::= LOAD_NAME expr CALL_FUNCTION_1 POP_TOP
		lc_body ::= LOAD_FAST expr CALL_FUNCTION_1 POP_TOP
		lc_body ::= LOAD_NAME expr LIST_APPEND
        lc_body ::= LOAD_FAST expr LIST_APPEND
		'''
		
	def p_augmented_assign(self, args):
		'''
		stmt ::= augassign1
		stmt ::= augassign2
		augassign1 ::= expr expr inplace_op designator
		augassign1 ::= expr expr inplace_op ROT_THREE STORE_SUBSCR
		augassign1 ::= expr expr inplace_op ROT_TWO   STORE_SLICE+0
		augassign1 ::= expr expr inplace_op ROT_THREE STORE_SLICE+1
		augassign1 ::= expr expr inplace_op ROT_THREE STORE_SLICE+2
		augassign1 ::= expr expr inplace_op ROT_FOUR  STORE_SLICE+3
		augassign2 ::= expr DUP_TOP LOAD_ATTR expr
				inplace_op ROT_TWO   STORE_ATTR

		inplace_op ::= INPLACE_ADD
		inplace_op ::= INPLACE_SUBTRACT
		inplace_op ::= INPLACE_MULTIPLY
		inplace_op ::= INPLACE_DIVIDE
		inplace_op ::= INPLACE_TRUE_DIVIDE
		inplace_op ::= INPLACE_FLOOR_DIVIDE
		inplace_op ::= INPLACE_MODULO
		inplace_op ::= INPLACE_POWER
		inplace_op ::= INPLACE_LSHIFT
		inplace_op ::= INPLACE_RSHIFT
		inplace_op ::= INPLACE_AND
		inplace_op ::= INPLACE_XOR
		inplace_op ::= INPLACE_OR 
		'''

	def p_assign(self, args):
		'''
		stmt ::= assign
		assign ::= expr DUP_TOP designList
		assign ::= expr designator
		'''

	def p_print(self, args):
		'''
		stmt ::= print_stmt
		stmt ::= print_stmt_nl
		stmt ::= print_nl_stmt
		print_stmt ::= expr PRINT_ITEM
		print_nl_stmt ::= PRINT_NEWLINE
		print_stmt_nl ::= print_stmt print_nl_stmt
		'''

	def p_print_to(self, args):
		'''
		stmt ::= print_to
		stmt ::= print_to_nl
		stmt ::= print_nl_to
		print_to ::= expr print_to_items POP_TOP
		print_to_nl ::= expr print_to_items PRINT_NEWLINE_TO
		print_nl_to ::= expr PRINT_NEWLINE_TO
		print_to_items ::= print_to_items print_to_item
		print_to_items ::= print_to_item
		print_to_item ::= DUP_TOP expr ROT_TWO PRINT_ITEM_TO
		'''
		# expr   print_to*   POP_TOP
		# expr { print_to* } PRINT_NEWLINE_TO

	def p_import15(self, args):
		'''
		stmt ::= importstmt
		stmt ::= importfrom

		importstmt ::= IMPORT_NAME STORE_FAST
		importstmt ::= IMPORT_NAME STORE_NAME

		importfrom ::= IMPORT_NAME importlist POP_TOP
		importlist ::= importlist IMPORT_FROM
		importlist ::= IMPORT_FROM
		'''

	def p_import20(self, args):
		'''
		stmt ::= importstmt2
		stmt ::= importfrom2
		stmt ::= importstar2

		importstmt2 ::= LOAD_CONST import_as
		importstar2 ::= LOAD_CONST IMPORT_NAME IMPORT_STAR

		importfrom2 ::= LOAD_CONST IMPORT_NAME importlist2 POP_TOP
		importlist2 ::= importlist2 import_as
		importlist2 ::= import_as
		import_as ::= IMPORT_NAME designator
		import_as ::= IMPORT_NAME LOAD_ATTR designator
		import_as ::= IMPORT_FROM designator
		'''

	def p_grammar(self, args):
		'''
		stmts ::= stmts stmt
		stmts ::= stmt

		stmts_opt ::= stmts
		stmts_opt ::= passstmt
		passstmt ::= 

		designList ::= designator designator
		designList ::= designator DUP_TOP designList

		designator ::= STORE_FAST
		designator ::= STORE_NAME
		designator ::= STORE_GLOBAL
		designator ::= STORE_DEREF
		designator ::= expr STORE_ATTR
		designator ::= expr STORE_SLICE+0
		designator ::= expr expr STORE_SLICE+1
		designator ::= expr expr STORE_SLICE+2
		designator ::= expr expr expr STORE_SLICE+3
		designator ::= store_subscr
		store_subscr ::= expr expr STORE_SUBSCR
		designator ::= unpack
		designator ::= unpack_list
		
		stmt ::= classdef
		stmt ::= call_stmt
		call_stmt ::= expr POP_TOP

		stmt ::= return_stmt
		return_stmt ::= expr RETURN_VALUE

		stmt ::= yield_stmt
		yield_stmt ::= expr YIELD_STMT
		yield_stmt ::= expr YIELD_VALUE

		stmt ::= break_stmt
		break_stmt ::= BREAK_LOOP
		
		stmt ::= continue_stmt
		continue_stmt ::= JUMP_ABSOLUTE
		continue_stmt ::= CONTINUE_LOOP
		
		stmt ::= raise_stmt
		raise_stmt ::= exprlist RAISE_VARARGS
		raise_stmt ::= nullexprlist RAISE_VARARGS
		
		stmt ::= exec_stmt
		exec_stmt ::= expr exprlist DUP_TOP EXEC_STMT
		exec_stmt ::= expr exprlist EXEC_STMT
		
		stmt ::= assert
		stmt ::= assert2
		stmt ::= assert3
		stmt ::= assert4
		stmt ::= ifstmt
		stmt ::= ifelsestmt
		stmt ::= whilestmt
		stmt ::= while1stmt
        stmt ::= while12stmt
		stmt ::= whileelsestmt
		stmt ::= while1elsestmt
		stmt ::= while12elsestmt
		stmt ::= forstmt
		stmt ::= forelsestmt
		stmt ::= trystmt
		stmt ::= tryfinallystmt
		
		stmt ::= del_stmt
		del_stmt ::= DELETE_FAST
		del_stmt ::= DELETE_NAME
		del_stmt ::= DELETE_GLOBAL
		del_stmt ::= expr DELETE_SLICE+0
		del_stmt ::= expr expr DELETE_SLICE+1
		del_stmt ::= expr expr DELETE_SLICE+2
		del_stmt ::= expr expr expr DELETE_SLICE+3
		del_stmt ::= delete_subscr
		delete_subscr ::= expr expr DELETE_SUBSCR
		del_stmt ::= expr DELETE_ATTR
		
		kwarg   ::= LOAD_CONST expr
		
		classdef ::= LOAD_CONST expr mkfunc
			        CALL_FUNCTION_0 BUILD_CLASS designator

		condjmp    ::= JUMP_IF_FALSE POP_TOP
		condjmp    ::= JUMP_IF_TRUE  POP_TOP

		assert ::= expr JUMP_IF_FALSE POP_TOP
				LOGIC_TEST_START expr JUMP_IF_TRUE POP_TOP
				LOGIC_TEST_START LOAD_GLOBAL RAISE_VARARGS
				LOGIC_TEST_END LOGIC_TEST_END POP_TOP
		assert2 ::= expr JUMP_IF_FALSE POP_TOP
				LOGIC_TEST_START expr JUMP_IF_TRUE POP_TOP
				LOGIC_TEST_START LOAD_GLOBAL expr RAISE_VARARGS
				LOGIC_TEST_END LOGIC_TEST_END POP_TOP
		assert3 ::= expr JUMP_IF_TRUE POP_TOP
				LOGIC_TEST_START LOAD_GLOBAL RAISE_VARARGS
				LOGIC_TEST_END POP_TOP
		assert4 ::= expr JUMP_IF_TRUE POP_TOP
				LOGIC_TEST_START LOAD_GLOBAL expr RAISE_VARARGS
				LOGIC_TEST_END POP_TOP

		_jump ::= JUMP_ABSOLUTE
		_jump ::= JUMP_FORWARD

		ifstmt ::= expr condjmp
                IF_THEN_START stmts_opt IF_THEN_END
                _jump POP_TOP IF_ELSE_START IF_ELSE_END

		ifelsestmt ::= expr condjmp
                IF_THEN_START stmts_opt IF_THEN_END
				_jump POP_TOP IF_ELSE_START stmts IF_ELSE_END

		trystmt ::= SETUP_EXCEPT TRY_START stmts_opt
				TRY_END POP_BLOCK _jump
				except_stmt

		try_end  ::= END_FINALLY TRY_ELSE_START TRY_ELSE_END
		try_end  ::= except_else
		except_else ::= END_FINALLY TRY_ELSE_START stmts TRY_ELSE_END

		except_stmt ::= except_cond except_stmt
		except_stmt ::= except_conds try_end
		except_stmt ::= except try_end
		except_stmt ::= try_end

		except_conds ::= except_cond except_conds
		except_conds ::= 

		except_cond ::= except_cond1
		except_cond ::= except_cond2
		except_cond1 ::= EXCEPT_START DUP_TOP expr COMPARE_OP
				JUMP_IF_FALSE
				POP_TOP POP_TOP POP_TOP POP_TOP
				stmts_opt EXCEPT_END _jump POP_TOP
		except_cond2 ::= EXCEPT_START DUP_TOP expr COMPARE_OP
				JUMP_IF_FALSE
				POP_TOP POP_TOP designator POP_TOP
				stmts_opt EXCEPT_END _jump POP_TOP
		except ::= EXCEPT_START POP_TOP POP_TOP POP_TOP
				stmts_opt EXCEPT_END _jump

		tryfinallystmt ::= SETUP_FINALLY stmts_opt
				POP_BLOCK LOAD_CONST
				stmts_opt END_FINALLY

        _while1test ::= _jump JUMP_IF_FALSE POP_TOP
        _while1test ::=

		whilestmt ::= SETUP_LOOP WHILE_START
				expr condjmp
				stmts_opt WHILE_END JUMP_ABSOLUTE
				WHILE_ELSE_START POP_TOP POP_BLOCK WHILE_ELSE_END

		while1stmt ::= SETUP_LOOP _while1test WHILE1_START
				stmts_opt WHILE1_END JUMP_ABSOLUTE
				WHILE1_ELSE_START POP_TOP POP_BLOCK WHILE1_ELSE_END

		while12stmt ::= SETUP_LOOP WHILE1_START
				_jump JUMP_IF_FALSE POP_TOP
				stmts_opt WHILE1_END JUMP_ABSOLUTE
				WHILE1_ELSE_START POP_TOP POP_BLOCK WHILE1_ELSE_END

		whileelsestmt ::= SETUP_LOOP WHILE_START
		       	expr condjmp
				stmts_opt WHILE_END JUMP_ABSOLUTE
				WHILE_ELSE_START POP_TOP POP_BLOCK
				stmts WHILE_ELSE_END

		while1elsestmt ::= SETUP_LOOP _while1test WHILE1_START
				stmts_opt WHILE1_END JUMP_ABSOLUTE
				WHILE1_ELSE_START POP_TOP POP_BLOCK
				stmts WHILE1_ELSE_END

		while12elsestmt ::= SETUP_LOOP WHILE1_START
				_jump JUMP_IF_FALSE POP_TOP
				stmts_opt WHILE1_END JUMP_ABSOLUTE
				WHILE1_ELSE_START POP_TOP POP_BLOCK
				stmts WHILE1_ELSE_END

		_for ::= GET_ITER FOR_START FOR_ITER
		_for ::= LOAD_CONST FOR_LOOP

		forstmt ::= SETUP_LOOP expr _for designator
				stmts_opt FOR_END JUMP_ABSOLUTE
				FOR_ELSE_START POP_BLOCK FOR_ELSE_END
		forelsestmt ::= SETUP_LOOP expr _for designator
				stmts_opt FOR_END JUMP_ABSOLUTE
				FOR_ELSE_START POP_BLOCK stmts FOR_ELSE_END

		'''

	def p_expr(self, args):
		'''
		expr ::= load_closure mklambda
		expr ::= mklambda
		expr ::= SET_LINENO
		expr ::= LOAD_FAST
		expr ::= LOAD_NAME
		expr ::= LOAD_CONST
		expr ::= LOAD_GLOBAL
		expr ::= LOAD_DEREF
		expr ::= LOAD_LOCALS
		expr ::= expr LOAD_ATTR
		expr ::= binary_expr
		expr ::= build_list
		
		binary_expr ::= expr expr binary_op
		binary_op ::= BINARY_ADD
		binary_op ::= BINARY_SUBTRACT
		binary_op ::= BINARY_MULTIPLY
		binary_op ::= BINARY_DIVIDE
		binary_op ::= BINARY_TRUE_DIVIDE
		binary_op ::= BINARY_FLOOR_DIVIDE
		binary_op ::= BINARY_MODULO
		binary_op ::= BINARY_LSHIFT
		binary_op ::= BINARY_RSHIFT
		binary_op ::= BINARY_AND
		binary_op ::= BINARY_OR
		binary_op ::= BINARY_XOR
		binary_op ::= BINARY_POWER

		expr ::= binary_subscr
		binary_subscr ::= expr expr BINARY_SUBSCR
		expr ::= expr expr DUP_TOPX_2 BINARY_SUBSCR
		expr ::= cmp
		expr ::= expr UNARY_POSITIVE
		expr ::= expr UNARY_NEGATIVE
		expr ::= expr UNARY_CONVERT
		expr ::= expr UNARY_INVERT
		expr ::= expr UNARY_NOT
		expr ::= mapexpr
		expr ::= expr SLICE+0
		expr ::= expr expr SLICE+1
		expr ::= expr expr SLICE+2
		expr ::= expr expr expr SLICE+3
		expr ::= expr DUP_TOP SLICE+0
		expr ::= expr expr DUP_TOPX_2 SLICE+1
		expr ::= expr expr DUP_TOPX_2 SLICE+2
		expr ::= expr expr expr DUP_TOPX_3 SLICE+3
		expr ::= and
		expr ::= and2
		expr ::= or
		or   ::= expr JUMP_IF_TRUE  POP_TOP LOGIC_TEST_START expr LOGIC_TEST_END
		and  ::= expr JUMP_IF_FALSE POP_TOP LOGIC_TEST_START expr LOGIC_TEST_END
		and2 ::= _jump JUMP_IF_FALSE POP_TOP LOGIC_TEST_START expr LOGIC_TEST_END

		cmp ::= cmp_list
		cmp ::= compare
		compare ::= expr expr COMPARE_OP
		cmp_list ::= expr cmp_list1 ROT_TWO IF_ELSE_START POP_TOP
				IF_ELSE_END
		cmp_list1 ::= expr DUP_TOP ROT_THREE
				COMPARE_OP JUMP_IF_FALSE POP_TOP
				cmp_list1
		cmp_list1 ::= expr DUP_TOP ROT_THREE
				COMPARE_OP JUMP_IF_FALSE POP_TOP
				IF_THEN_START cmp_list1
		cmp_list1 ::= expr DUP_TOP ROT_THREE
				COMPARE_OP JUMP_IF_FALSE POP_TOP
				IF_THEN_START cmp_list2
		cmp_list1 ::= expr DUP_TOP ROT_THREE
				COMPARE_OP JUMP_IF_FALSE POP_TOP
				cmp_list2
		cmp_list2 ::= expr COMPARE_OP IF_THEN_END JUMP_FORWARD
		mapexpr ::= BUILD_MAP kvlist

		kvlist ::= kvlist kv
		kvlist ::= kvlist kv2
		kvlist ::=

		kv ::= DUP_TOP expr ROT_TWO expr STORE_SUBSCR
		kv2 ::= DUP_TOP expr expr ROT_THREE STORE_SUBSCR

		exprlist ::= exprlist expr
		exprlist ::= expr

		nullexprlist ::=
		'''

	def nonterminal(self, nt, args):
		collect = ('stmts', 'exprlist', 'kvlist')

		if nt in collect and len(args) > 1:
			#
			#  Collect iterated thingies together.
			#
			rv = args[0]
			rv.append(args[1])
		else:
			rv = GenericASTBuilder.nonterminal(self, nt, args)
		return rv

	def __ambiguity(self, children):
		# only for debugging! to be removed hG/2000-10-15
		print children
		return GenericASTBuilder.ambiguity(self, children)

	def resolve(self, list):
		if len(list) == 2 and 'funcdef' in list and 'assign' in list:
			return 'funcdef'
		#print >> sys.stderr, 'resolve', str(list)
		return GenericASTBuilder.resolve(self, list)

nop = lambda self, args: None

p = Parser()

def parse(tokens, customize):
	#
	#  Special handling for opcodes that take a variable number
	#  of arguments -- we add a new rule for each:
	#
	#	expr ::= {expr}^n BUILD_LIST_n
	#	expr ::= {expr}^n BUILD_TUPLE_n
	#	expr ::= {expr}^n BUILD_SLICE_n
	#	unpack_list ::= UNPACK_LIST {expr}^n
	#	unpack ::= UNPACK_TUPLE {expr}^n
	#	unpack ::= UNPACK_SEQEUENE {expr}^n
	#	mkfunc ::= {expr}^n LOAD_CONST MAKE_FUNCTION_n
	#	mkfunc ::= {expr}^n load_closure LOAD_CONST MAKE_FUNCTION_n
	#	expr ::= expr {expr}^n CALL_FUNCTION_n
	#	expr ::= expr {expr}^n CALL_FUNCTION_VAR_n POP_TOP
	#	expr ::= expr {expr}^n CALL_FUNCTION_VAR_KW_n POP_TOP
	#	expr ::= expr {expr}^n CALL_FUNCTION_KW_n POP_TOP
	#
	global p
	for k, v in customize.items():
		# avoid adding the same rule twice to this parser
		if p.customized.has_key(k):
			continue
		p.customized[k] = None

		#nop = lambda self, args: None
		op = k[:string.rfind(k, '_')]
		if op in ('BUILD_LIST', 'BUILD_TUPLE'):
			rule = 'build_list ::= ' + 'expr '*v + k
		elif op == 'BUILD_SLICE':
			rule = 'expr ::= ' + 'expr '*v + k
		elif op in ('UNPACK_TUPLE', 'UNPACK_SEQUENCE'):
			rule = 'unpack ::= ' + k + ' designator'*v
		elif op == 'UNPACK_LIST':
			rule = 'unpack_list ::= ' + k + ' designator'*v
		elif op == 'DUP_TOPX':
			# no need to add a rule
			continue
			#rule = 'dup_topx ::= ' + 'expr '*v + k
		elif op == 'MAKE_FUNCTION':
			p.addRule('mklambda ::= %s LOAD_LAMBDA %s' %
				  ('expr '*v, k), nop)
			rule = 'mkfunc ::= %s LOAD_CONST %s' % ('expr '*v, k)
		elif op == 'MAKE_CLOSURE':
			p.addRule('mklambda ::= %s load_closure LOAD_LAMBDA %s' %
				  ('expr '*v, k), nop)
			rule = 'mkfunc ::= %s load_closure LOAD_CONST %s' % ('expr '*v, k)
		elif op in ('CALL_FUNCTION', 'CALL_FUNCTION_VAR',
			    'CALL_FUNCTION_VAR_KW', 'CALL_FUNCTION_KW'):
			na = (v & 0xff)           # positional parameters
			nk = (v >> 8) & 0xff      # keyword parameters
			# number of apply equiv arguments:
			nak = ( len(op)-len('CALL_FUNCTION') ) / 3
			rule = 'expr ::= expr ' + 'expr '*na + 'kwarg '*nk \
			       + 'expr ' * nak + k
		else:
			raise 'unknown customize token %s' % k
		p.addRule(rule, nop)
	ast = p.parse(tokens)
	#p.cleanup()
	return ast

# local variables:
# tab-width: 4


decompyle/magics.py
import struct

__all__ = ['magics', 'versions']

def __build_magic(magic):
    return struct.pack('Hcc', magic, '\r', '\n')

def __by_version(magics):
    by_version = {}
    for m, v in magics.items():
        by_version[v] = m
    return by_version

versions = {
    # magic, version
    __build_magic(20121): '1.5',
    __build_magic(50428): '1.6',
    __build_magic(50823): '2.0',
    __build_magic(60202): '2.1',
    __build_magic(60717): '2.2',
    __build_magic(62011): '2.3',
    __build_magic(62061): '2.4',
}

magics = __by_version(versions)

def __show(text, magic):
    print text, struct.unpack('BBBB', magic), \
          struct.unpack('HBB', magic)

def test():
    import imp
    magic_20 = by_version['2.0']
    current = imp.get_magic()
    current_version = magics[current]
    magic_current = by_version[ current_version ]
    print type(magic_20), len(magic_20), repr(magic_20)
    print
    print 'This Python interpreter has version', current_version
    __show('imp.get_magic():\t', current),
    __show('magic[current_version]:\t', magic_current)
    __show('magic_20:\t\t', magic_20)
    
if __name__ == '__main__':
    test()


decompyle/marshal_files.py
import magics

__all__ = ['by_version', 'by_magic']

by_version = {
    '1.5': 'marshal_20',
    '1.6': 'marshal_20',
    '2.0': 'marshal_20',
    '2.1': 'marshal_22',
    '2.2': 'marshal_22',
    '2.3': 'marshal_23',
    '2.4': 'marshal_24',
}

by_magic = dict( [ (mag, by_version[ver])
                   for mag, ver in magics.versions.iteritems() ] )

def import_(module=None,version=None,magic=None):
    if module:    pass
    elif version: module = by_version[version]
    elif magic:   module = by_magic[magic]
    else:
        raise 'at least one argument is required'
    from __builtin__ import __import__
    if module == 'marshal':
        # use current version's 'marshal' module
        return __import__('marshal', globals(), locals())
    else:
        return __import__('decompyle.%s' % module, globals(),
                          locals(), 'decompyle')



decompyle/dis_files.py
import magics

__all__ = ['by_version', 'by_magic']

_fallback = {
    'EXTENDED_ARG': None,
    'hasfree': [],
    }

class dis(object):
    def __init__(self, version, module):
        self._version = version
        from __builtin__ import __import__
        self._module = __import__('decompyle.%s' % module, globals(),
                                  locals(), 'decompyle')

    def __getattr__(self, attr):
        try:
            val = self._module.__dict__[attr]
        except KeyError, e:
            if _fallback.has_key(attr):
                val = _fallback[attr]
            else:
                raise e
        return val

by_version = {
    '1.5': dis('1.5', 'dis_15'),
    '1.6': dis('1.6', 'dis_16'),
    '2.0': dis('2.0', 'dis_20'),
    '2.1': dis('2.1', 'dis_21'),
    '2.2': dis('2.2', 'dis_22'),
    '2.3': dis('2.3', 'dis_23'),
    '2.4': dis('2.4', 'dis_24'),
}

by_magic = dict( [ (mag, by_version[ver])
                   for mag, ver in magics.versions.iteritems() ] )

if __name__ == '__main__':
    for m, ver in by_magic.items():
        magics.__show(ver, m)
    print by_version['2.2'].hasjrel
